### ISSUE TITLE
2.0.4.13 DB bug report code keeps original FileLookup Path intact (privacy regression)

### ISSUE THREAD

**Issue by ts678** [duplicati/duplicati#3790]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.4.17
- **Operating system**: Windows 10 Version 1809
- **Backend**: local folder

## Description

The DB bug report has long scrubbed sensitive information such as source pathnames. Various forum posts and [documentation](https://duplicati.readthedocs.io/en/latest/03-using-the-graphical-user-interface/#creating-a-bug-report) describe this. [2.0.4.13 canary](https://github.com/duplicati/duplicati/releases/tag/v2.0.4.13-2.0.4.13_canary_2019-01-29) code for the new path storage scheme (maybe inadvertently) leaves the FileLookup table intact including its Path. This could not be seen until [2.0.4.17 canary](https://github.com/duplicati/duplicati/releases/tag/v2.0.4.17-2.0.4.17_canary_2019-04-11) fixed ["Create bug report ..." results in syntax error #3725](https://github.com/duplicati/duplicati/issues/3725). I'm not sure which tables are most helpful. The new code creates one it calls FixedFile, and drops others. Should it drop or fix FileLookup?

https://github.com/duplicati/duplicati/blob/b85b817d674cc63e0bec4c6201b1ddf01be71bac/Duplicati/Library/Main/Database/LocalBugReportDatabase.cs#L68-L70

## Steps to reproduce

1. Backup
2. "Create bug report"
3. Look at the FileLookup table with a viewer such as DB Browser for SQLite

- **Actual result**:
Path is sitting there readable.
- **Expected result**:
Obfuscated path or no table.

## Screenshots

![image](https://user-images.githubusercontent.com/40771998/58830208-cb9dd880-8617-11e9-83a1-dec8875f7e9e.png)

**Comment by ts678** [comment]
[Feature/fix path storage2 #3468](https://github.com/duplicati/duplicati/pull/3468) might be the pull request. There are some side issues we could get into, such as the future of DB bug reports. People haven't been very willing to post them (although some have), and this makes it hard to dig into field failures for the problems that developers can't reproduce themselves. Regarding deletion, a VACUUM would be more thorough, otherwise deleted information remains. Tools can dig data out.

**Comment by ts678** [comment]
Rather than open a different issue on VACUUM (which is probably an issue dating back to the earliest code):

[Re: [sqlite] Parsing sqlitedb file.](https://www.mail-archive.com/sqlite-users@mailinglists.sqlite.org/msg71542.html) describes the problem, and the SQLite leader's advice, which is linked below:

[PRAGMA schema.secure_delete;](https://www.sqlite.org/pragma.html#pragma_secure_delete) which says VACUUM should work too. That fits the bug report usage better.

Although web search can find many tools that _claim_ to recover deleted data, the sqlite mailing list discussion (and other resources) claim a proper recovery is iffy. From a privacy point of view though, the [scrubbing](https://github.com/duplicati/duplicati/issues/2298#issuecomment-276895799) that Duplicati does leaves things like pathnames behind, easily seen by looking into the file, e.g. with a hex editor.

2.0.4.13 does expand the privacy leak to be obvious with the tools one _normally_ uses, e.g. an SQLite browser.

**Comment by BlueBlock** [comment]
I'm thinking the create-bug-report could default to heavy scrubbing but also give the option to include the private data if the user chooses if it were to help debug a problem.

Might you be able to provide a list of the tables and/or fields you see as impacting privacy?

**Comment by ts678** [comment]
> Might you be able to provide a list of the tables and/or fields you see as impacting privacy?

Looking at the [2.0.4.22 canary schema](https://github.com/duplicati/duplicati/blob/v2.0.4.22-2.0.4.22_canary_2019-06-30/Duplicati/Library/Main/Database/Database%20schema/Schema.sql) I think it's:

PathPrefix table Prefix
FileLookup table Path
File view Path (the view is made from FileLookup and PathPrefix which replaced File table for lookups)
LogData table Message

Earlier Duplicati versions such as 2.0.4.5 beta cleaned up File table and LogData table. It looks like the replacement plan creates a FixedFile table to replace the File view, and cleans it like it did the old File.

PathPrefix table cleaning is not needed because it gets deleted. Maybe delete FileLookup table here?

https://github.com/duplicati/duplicati/blob/09be92c59e4ea1cffb834884d6b6433a1aeb1fef/Duplicati/Library/Main/Database/LocalBugReportDatabase.cs#L70

For reference and comparison, 2.0.4.5 Fix method is [here](https://github.com/duplicati/duplicati/blob/79d839546b238be79967bf8532267bfab2b08aa6/Duplicati/Library/Main/Database/LocalBugReportDatabase.cs#L35).

One can infer that the main target of Fix is pathnames from the below, and maybe it's a spot to always vacuum (maybe controversial because it might take awhile, but this is a bug report and not a backup):

https://github.com/duplicati/duplicati/blob/09be92c59e4ea1cffb834884d6b6433a1aeb1fef/Duplicati/Library/Main/Operation/CreateBugReportHandler.cs#L59-L72

One can infer that logs (being somewhat free-form as opposed to SQL tables) are less surely scrubbed:

https://github.com/duplicati/duplicati/blob/09be92c59e4ea1cffb834884d6b6433a1aeb1fef/Duplicati/CommandLine/Commands.cs#L786-L788

In other things that worry me less and aren't regressions:

RemoteOperation table Data for a list operation might see a user-made name, but I think it's unlikely. Filtering on [--prefix](https://duplicati.readthedocs.io/en/latest/06-advanced-options/#prefix) could reduce `list` of irrelevant filenames, if someone thinks that's a better plan.

Configuration table Passphrase seems to me to have no debug benefit. I think it's a salted hash of the passphrase, but it'd seem better security practice to not have it there at all unless it has some benefit.

**Comment by drwtsn32x** [comment]
Is dropping the FileLookup table the right solution? It is not needed for debugging purposes?

I agree that Configuration table 'passphrase' value should be removed.

I think LogData table, Messages field could be beneficial in debugging as it would log specific error codes and it includes other information like exact version used.  I don't see any sensitive data there in my test database.

Thoughts?

**Comment by BlueBlock** [comment]
I'm thinking the best solution might just be to offer the user choices of which types of data they want to include/exclude in the bug report. 

Issues really need to be reproducible so, if needed, the user could later be requested to reproduce the issue and generate a new bug report and provide the additional information.

Enhancement? Just as there is the usage report, it would be really nice to have errors and exceptions automatically reported but this would have to be anatomized and scrubbed data only.

**Comment by drwtsn32x** [comment]
> I'm thinking the best solution might just be to offer the user choices of which types of data they want to include/exclude in the bug report.

Not a bad idea.. I get where you're coming from. Then again I don't see the downside to sanitizing file/path data.  Don't delete it, just change it (like the current sanitize operation already does when it converts the regular File view to a new FixedFile table.

This should not diminish debugging ability at all, right?  Same with clearing the Configuration.Passphrase value - don't see how that would hurt debugging.

Dropping tables entirely may be a different story though. The code already drops PathPrefix, so presumably the devs decided that is not needed for debugging.  Otherwise we could just sanitize it.

Main question is - would it hurt debugging to drop FileLookup table?  If it would, then can we sanitize it instead?

To get back to your point though - the user does already have the option of providing the original database to the devs if they do not have any privacy concerns.  (It's just not an option provided when you use the bug report feature... they'd have to manually zip the database.)

**Comment by ts678** [comment]
TL;DR My main goal in filing this issue was to fix a regression. There's certainly enhancement possible.

For LogData, possibly the thought was that certain error messages have filenames. I haven't looked at the analysis method closely, but I think I've seen forum DB bug reports with lots of LogData removed.

Dropping the FileLookup table seemed the other half of dropping the PathPrefix table. The two replace what used to be the File table. Currently the File "table" is actually an on-the-fly view from those tables.

Debugging might go better if both tables were kept but sanitized, however sanitization was (I believe) previously done in the File table by mapping distinct paths to distinct fake pathnames. With File Path now built from a PathPrefix table Prefix and a FileLookup table Path, it seems possible that one could build the same effective File view Path multiple ways, however it might be OK if the privacy-enhanced versions get a little weird. If you can sanitize just FileLookup, that might work, but arguably PathPrefix should have been done similarly, yet was simply dropped instead. Dropping FileLookup matches that.

Debugging the new table scheme seems a rarer need than, say, trying to debug the typical broken DB situation. I see your point, but I'm also trying to get privacy fast before some release window closes. If need be, it can presumably be undone later and better (maybe even solving my uniqueness concerns).

[Feature/fix path storage2 #3468](https://github.com/duplicati/duplicati/pull/3468) is likely where code came in, but privacy vs. debug wasn't discussed...

**Comment by drwtsn32x** [comment]
> but I think I've seen forum DB bug reports with lots of LogData removed.

Yep there is code already in the sanitization routine that appears to remove path information from LogData.Message and LogData.Exception.

> Dropping the FileLookup table seemed the other half of dropping the PathPrefix table. The two replace what used to be the File table. Currently the File "table" is actually an on-the-fly view from those tables.

I just came to that conclusion after looking more closely at the code.  Since the new sanitized FixedFile table is not a view but a real table, neither FileLookup nor PathPrefix are needed or are useful it seems.

I will submit a pull request where PathPrefix is dropped and the Configuration.passphrase value gets erased.

**Comment by Pectojin** [comment]
Fixed by #3842
